core/store/src/trie/trie_storage.rs

```diff
@@ -9,34 +9,144 @@ use crate::trie::POISONED_LOCK_ERR;
 use crate::{ColState, StorageError, Store};
 use lru::LruCache;
 use near_primitives::shard_layout::ShardUId;
+use near_primitives::types::TrieCacheState;
 use std::cell::RefCell;
 use std::io::ErrorKind;
 
 #[derive(Clone)]
-pub struct TrieCache(Arc<Mutex<LruCache<CryptoHash, Vec<u8>>>>);
+pub struct SyncTrieCache(Arc<Mutex<TrieCache>>);
+
+/// Cache for Trie nodes.
+/// Maps hash of the encoded node to the encoded node bytes.
+/// Note that "node" here means both the trie node and the value stored in trie.
+/// TODO (#5920): enable chunk nodes caching in Runtime::apply.
+pub(crate) struct TrieCache {
+    pub(crate) cache_state: TrieCacheState,
+    shard_cache: LruCache<CryptoHash, Vec<u8>>,
```

Comment:
later we should check how much faster would it be to use Arc<Vec<u8>> here (and 1 line below)